<!DOCTYPE html>
<html>
  <head>
    <title>radmixture教程</title>
    <meta charset="utf-8">
    <meta name="author" content="边蓓蕾" />
    <script src="libs/htmlwidgets/htmlwidgets.js"></script>
    <link href="libs/plotlyjs/plotly-htmlwidgets.css" rel="stylesheet" />
    <script src="libs/plotlyjs/plotly-latest.min.js"></script>
    <script src="libs/plotly-binding/plotly.js"></script>
    <link rel="stylesheet" href="zh-CN.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# radmixture教程
### 边蓓蕾
### 2017/03/05

---




# 什么是radmixture?

- radmixture是一个提供祖源计算服务的开源R包。
- 如果您有如下格式的原始数据，可以用此包提供的数据集得到基于不同参考人群计算出来的结果。


```
#&gt;          rsid chromosome position genotype
#&gt; 1  rs12564807          1   734462       AA
#&gt; 2   rs3131972          1   752721       AG
#&gt; 3 rs148828841          1   760998       CC
#&gt; 4  rs12124819          1   776546       AG
#&gt; 5 rs115093905          1   787173       GG
#&gt; 6  rs11240777          1   798959       AG
```

- 本包目前有一份K13数据集和一份E11数据集可以用来计算祖源。
- 除了面向个人基因组用户，也面向科研用户。如果您有自己的reference data，可以用有监督或无监督模型计算祖源成份。
---
### 使用方法

可以通过以下两种方法安装此包：

- 从**GitHub**上安装最新的开发版本：


```r
devtools::install_github("wegene-llc/radmixture")
```

然后加载此包：


```r
library(radmixture)
```

- 再等等就可以从CRAN上安装稳定版本了。
---
class: center, middle
background-image: url(https://s3-us-west-2.amazonaws.com/s.cdpn.io/44216/stars-img-src.svg)

&lt;h1 style="color:white; font-size: 40px;"&gt;&lt;strong&gt;和我一起在浩瀚的宇宙里自我探索&lt;/strong&gt;&lt;/h1&gt;

---
class: center, middle, inverse

# 使用公共数据集计算你的祖源成份
### 针对消费者

---
### 准备好你的数据集

- 首先，你一定要有一份长这样子的数据集


```
#&gt;          rsid chromosome position genotype
#&gt; 1  rs12564807          1   734462       AA
#&gt; 2   rs3131972          1   752721       AG
#&gt; 3 rs148828841          1   760998       CC
#&gt; 4  rs12124819          1   776546       AG
#&gt; 5 rs115093905          1   787173       GG
#&gt; 6  rs11240777          1   798959       AG
```
- 接下来，你只需要两个函数帮你完成这个有意思的事情。
    * `tfrdpub()`:帮你将原始数据转换为radmixture可以用来计算的格式。
    * `fFixQN()`:用于计算结果。
---
举个栗子：

第一步:将你的原始数据读入R，此时它是个data frame(如何读入取决于你拿到的数据格式，比如.txt或.csv)

```r
genotype &lt;- read.table(file = 'path/to/file')
genotype &lt;- read.csv(file = 'path/to/file')
```
第二步：利用本包内置的数据集转换你的原始数据，只需要将第一个参数改为刚刚读入的数据集，后面3个参数不动。

```r
# K13
res &lt;- tfrdpub(genotype, 13, global13.alleles, global13.13.F)
# E11
res &lt;- tfrdpub(genotype, 11, e11.alleles, e11.11.F)
```
第三步：计算结果

```r
ances &lt;- fFixQN(res$g, res$q, res$f, 1e-4, 'BR')
```
第四步：`ances$q`就是你的K13或E11的结果了。

---
class: center, middle, inverse
# 使用自己的reference panel计算祖源
### 针对科研用户

---
### 准备好你的数据集
- 首先，你要有一个genotype的矩阵，为计算做准备。

---
class: center, middle, inverse

# 文末彩蛋

.footnote[~~纯属个人喜好，不代表我司观点~~]
---
### 一些可视化的小建议

细心的你会发现，本包并没有提供任何可视化函数供大家玩耍，那是因为可以用来玩耍可视化的包足够多。

- plotly


```r
library(plotly)
data &lt;- tfrdpub(genotype, 13, global13.alleles, global13.13.F)
res &lt;- f.fix.qn(data$g, data$q, data$f, tol = 1e-4, method = "BR", pubdata = "K13")
label &lt;- c('Siberian', 'Amerindian', 'West_African', 'Palaeo_African',
                            'Southwest_Asian', 'East_Asian', 'Mediterranean','Australasian',
                            'Arctic', 'West_Asian', 'North_European', 'South_Asian',
           'East_African')
m &lt;- list(l = 30, r = 250, b = 10, t = 200)
            plot_ly(labels = label, values = as.numeric(res$q), 
                    domain = list(x = c(0, 2), y = c(0, 3))) %&gt;%
                add_pie(hole = 0.6) %&gt;%
                layout(
                    xaxis = list(showgrid = F, zeroline = F, showticklabels = F),
                    yaxis = list(showgrid = F, zeroline = F, showticklabels = F),
                    plot_bgcolor = "transparent",
                    paper_bgcolor = "transparent",
                    width = 500, height = 400, margin = m)
```

---
以上代码可产生如下效果：

<div id="htmlwidget-8e3c3d30c2f6a39dcb8f" style="width:504px;height:504px;" class="plotly html-widget"></div>
<script type="application/json" data-for="htmlwidget-8e3c3d30c2f6a39dcb8f">{"x":{"layout":{"width":500,"height":400,"margin":{"b":10,"l":30,"t":200,"r":250},"xaxis":{"domain":[0,1],"showgrid":false,"zeroline":false,"showticklabels":false},"yaxis":{"domain":[0,1],"showgrid":false,"zeroline":false,"showticklabels":false},"plot_bgcolor":"transparent","paper_bgcolor":"transparent"},"source":"A","config":{"modeBarButtonsToAdd":[{"name":"Collaborate","icon":{"width":1000,"ascent":500,"descent":-50,"path":"M487 375c7-10 9-23 5-36l-79-259c-3-12-11-23-22-31-11-8-22-12-35-12l-263 0c-15 0-29 5-43 15-13 10-23 23-28 37-5 13-5 25-1 37 0 0 0 3 1 7 1 5 1 8 1 11 0 2 0 4-1 6 0 3-1 5-1 6 1 2 2 4 3 6 1 2 2 4 4 6 2 3 4 5 5 7 5 7 9 16 13 26 4 10 7 19 9 26 0 2 0 5 0 9-1 4-1 6 0 8 0 2 2 5 4 8 3 3 5 5 5 7 4 6 8 15 12 26 4 11 7 19 7 26 1 1 0 4 0 9-1 4-1 7 0 8 1 2 3 5 6 8 4 4 6 6 6 7 4 5 8 13 13 24 4 11 7 20 7 28 1 1 0 4 0 7-1 3-1 6-1 7 0 2 1 4 3 6 1 1 3 4 5 6 2 3 3 5 5 6 1 2 3 5 4 9 2 3 3 7 5 10 1 3 2 6 4 10 2 4 4 7 6 9 2 3 4 5 7 7 3 2 7 3 11 3 3 0 8 0 13-1l0-1c7 2 12 2 14 2l218 0c14 0 25-5 32-16 8-10 10-23 6-37l-79-259c-7-22-13-37-20-43-7-7-19-10-37-10l-248 0c-5 0-9-2-11-5-2-3-2-7 0-12 4-13 18-20 41-20l264 0c5 0 10 2 16 5 5 3 8 6 10 11l85 282c2 5 2 10 2 17 7-3 13-7 17-13z m-304 0c-1-3-1-5 0-7 1-1 3-2 6-2l174 0c2 0 4 1 7 2 2 2 4 4 5 7l6 18c0 3 0 5-1 7-1 1-3 2-6 2l-173 0c-3 0-5-1-8-2-2-2-4-4-4-7z m-24-73c-1-3-1-5 0-7 2-2 3-2 6-2l174 0c2 0 5 0 7 2 3 2 4 4 5 7l6 18c1 2 0 5-1 6-1 2-3 3-5 3l-174 0c-3 0-5-1-7-3-3-1-4-4-5-6z"},"click":"function(gd) { \n        // is this being viewed in RStudio?\n        if (location.search == '?viewer_pane=1') {\n          alert('To learn about plotly for collaboration, visit:\\n https://cpsievert.github.io/plotly_book/plot-ly-for-collaboration.html');\n        } else {\n          window.open('https://cpsievert.github.io/plotly_book/plot-ly-for-collaboration.html', '_blank');\n        }\n      }"}],"modeBarButtonsToRemove":["sendDataToCloud"]},"data":[{"labels":["Siberian","Amerindian","West_African","Palaeo_African","Southwest_Asian","East_Asian","Mediterranean","Australasian","Arctic","West_Asian","North_European","South_Asian","East_African"],"values":[1e-05,0.00691,1e-05,1e-05,0.00554,1e-05,0.34674,0.00467,1e-05,0.09733,0.51691,0.01875,0.0031],"domain":{"x":[0,2],"y":[0,3]},"type":"pie","hole":0.6,"marker":{"fillcolor":"rgba(31,119,180,1)","color":"rgba(31,119,180,1)","line":{"color":"transparent"}}}],"base_url":"https://plot.ly"},"evals":["config.modeBarButtonsToAdd.0.click"],"jsHooks":[]}</script>
---
- 地图可视化

如果你对一些地图的js库和插件比较熟悉的话，让我先来膜拜你一下。这里先用一个简单的DataMaps库来帮你开启大门。

.center[![](/Users/bianbeilei/Desktop/map.png)]

先把压缩版的D3.js, Topojson, Datamaps.js下载到本地，然后在html里引用。把下一页一句代码
`var res = [0.02069,0.91506,0.0139,0.05035];` 改成你自己的结果. 用 `rjson::toJSON(res$q)`可以拿到这个json对象。

下面这段js代码是有点烦躁，毕竟我们都不熟悉js T_T，那我们换一个在R里面玩耍的好了。

.footnote[[1]~~一不小心就说出了htmlwidgets的原理~~]

---
```html
&lt;!DOCTYPE html&gt;
&lt;meta charset="utf-8"&gt;
  &lt;script src="d3.v3.min.js"&gt;&lt;/script&gt;
  &lt;script src="topojson.v1.min.js"&gt;&lt;/script&gt;
  &lt;script src="datamaps.world.min.js"&gt;&lt;/script&gt;
  &lt;div id="container" style="position: relative; width: 250px; max-height: 25px; float: left; padding:5px"&gt;&lt;/div&gt;

     &lt;script&gt;
     var res = [0.02069,0.91506,0.0139,0.05035];
     var percent = res.map(function(el) {
         return el * 100;
     });
     percent = percent.map(String);
     percent = percent.map(function(el) {
         return el.concat('%');
     });
      var map = new Datamap({
        scope: 'world',
        projection: 'mercator',
        element: document.getElementById('container'),
        height: 180,
        fills: {
          defaultFill: '#FFCCFF',
          'African': '#000000',
          'Amerindian': '#CCCCFF',
          'European': '#FFCC99',
          'Asian': '#CCFFFF'
        },
```
---
```html
        geographyConfig: {
            highlightOnHover: false,
            popupOnHover: false,
            borderColor: '#FFCCFF'
        }
      });
      map.bubbles([
       {name: 'European: ' + percent[0], latitude: 55, longitude: 21.27, radius: 18, fillKey: 'European'},
       {name: 'Asian: ' + percent[1], latitude: 30.97, longitude: 103.18, radius: 24, fillKey: 'Asian'},
       {name: 'African: ' + percent[2], latitude: -3.26, longitude: 18.73, radius: 24, fillkey: 'African'},
       {name: 'Amerindian: ' + percent[3], latitude: 21.32, longitude: -84.32, radius: 36, fillKey: 'Amerindian'}
     ], {
       popupTemplate: function(geo, data) {
         return "&lt;div class='hoverinfo'&gt;" + data.name.bold().italics().fontsize(2).fontcolor("green") + "&lt;/div&gt;";
       }
     });
     &lt;/script&gt;
```

---
class: center, middle, inverse

# Leaflet + mapview + sp

---
class: center, middle
    </textarea>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true
});
if (window.HTMLWidgets) slideshow.on('showSlide', function (slide) {setTimeout(function() {window.dispatchEvent(new Event('resize'));}, 100)});</script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  tex2jax: {
    skipTags: ['script', 'noscript', 'style', 'textarea', 'pre']
  }
});
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
